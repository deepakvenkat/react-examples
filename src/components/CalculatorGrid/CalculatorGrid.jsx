import React from 'react';
import PropTypes from 'prop-types';
import { Grid } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';

import CalculatorButton from '../CalculatorButton';
import ResultArea from '../ResultArea';
import ClearButton from '../ClearButton';

import useResultValue from './hooks/useResultValue';

const styles = {
  root: {},
};

const buttonCbGenerator = (value, dispatch) => () => {
  dispatch({
    type: value,
  });
};

export const CalculatorRow = ({
  row,
  dispatch
}) => {
  const rowDom = row.map((element) => {
    const onClickFunction = buttonCbGenerator(element, dispatch);
    if (typeof element === 'string'){
      if (element === 'clear') {
        return <ClearButton onClick={onClickFunction} key="clear"/>;
      } else {
        return (
          <CalculatorButton
            type="operator"
            text={element}
            onClick={onClickFunction}
            key={element}
          />
        )
      }
    } else if (typeof element === 'number'){
      return (
        <CalculatorButton
          type="number"
          text={`${element}`}
          onClick={onClickFunction}
          key={element}
        />
      )
    } else {
      throw new Error('can\'t handle element');
    }
  });
  return rowDom;
};
const layout = [
  ['clear', 'AC'],
  [7, 8, 9, '+'],
  [4, 5, 6, '-'],
  [1, 2, 3, '*'],
  [0, '.', '=', '/'],
];

const CalculatorGrid = () => {
  const {
    result: resultValue,
    dispatch,
  } = useResultValue();
  return (
    <Grid>
    <ResultArea
      value={resultValue}
    />
    { layout.map((row) => (
      <div>
        <CalculatorRow
          row={row}
          key={row[0]}
          dispatch={dispatch}
        />
      </div>
    ))}
    </Grid>
  );
};

CalculatorGrid.propTypes = {
  classes: PropTypes.shape({}).isRequired,
}
export default withStyles(styles)(CalculatorGrid);
