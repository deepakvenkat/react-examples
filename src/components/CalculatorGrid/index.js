import CalculatorGrid from './CalculatorGrid.jsx';

export default CalculatorGrid;
