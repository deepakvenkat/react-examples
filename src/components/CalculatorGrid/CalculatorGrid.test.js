import React from 'react';
import { render } from '@testing-library/react';
import CalculatorGrid from './CalculatorGrid';

test('renders the resultarea', () => {
  const { asFragment } = render(
    <CalculatorGrid />
  );
  expect(asFragment()).toMatchSnapshot();
});
