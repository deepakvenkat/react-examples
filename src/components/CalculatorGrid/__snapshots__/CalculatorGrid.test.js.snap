// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the resultarea 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root"
  >
    <div
      class="MuiInputBase-root ResultArea-root-2"
    >
      <input
        class="MuiInputBase-input"
        type="text"
        value=""
      />
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-5 ClearButton-root-3 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          clear
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-6 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          /
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-7 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          7
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-8 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          8
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-9 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          9
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-10 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          -
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-11 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          4
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-12 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          5
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-13 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          6
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-14 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          +
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-15 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          1
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-16 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          2
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-17 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          3
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained CalculatorButton-root-4 CalculatorButton-root-18 MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          =
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
